### PR TITLE
feat(streaming): split chunks in hash join

### PR DESCRIPTION
**Pull request by pleiadesian**
## What's changed and what's your intention?

As title. Also, `StreamChunkBuilder` is completely refactored to support capacity-limited chunk generation.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
#1891

### CODE REVIEW

File: src/stream/src/common/builder.rs [hunk 1]
```diff
@@ -52,15 +52,22 @@ pub struct StreamChunkBuilder {
...
 impl StreamChunkBuilder {
     pub fn new(
         capacity: usize,
         data_types: &[DataType],
         update_start_pos: usize,
         matched_start_pos: usize,
     ) -> Result<Self> {
-        // TODO: assert `capacity != 0`
-        let ops = Vec::with_capacity(capacity);
+        assert!(capacity > 1);
```
Comment (TennyZhuang) [inline src/stream/src/common/builder.rs:68]: `>= 1`?

File: src/stream/src/common/builder.rs [hunk 2]
```diff
@@ -52,15 +52,22 @@ pub struct StreamChunkBuilder {
...
+        let capacity = capacity - 1;
```
Comment (TennyZhuang) [inline src/stream/src/common/builder.rs:69]: Why `-1` here? Is there any comment?

Comment (pleiadesian) [reply src/stream/src/common/builder.rs:71]: Added comment for this.

For example, when `capacity = 3`, we generate a chunk when `size = capacity - 1 = 2`.
However, when the last op is `UpdateDelete`, we leave a space for the upcoming `UpdateInsert`, after which we can generate a chunk without splitting `UpdateDelete/UpdateInsert` or violating capacity requirement.
```
capacity = 3:
------------------------------------
Delete -> Delete (Generate a chunk)
------------------------------------
Insert -> UpdateDelete -> ? (Do not generate a chunk)
Insert -> UpdateDelete -> UpdateInsert (Generate a chunk)
```

**Comment by TennyZhuang** [comment]
Huge conflicts with #2387, and I'd prefer merge it first. I will resolve the conflicts after it merged.

**Comment by jon-chuang** [comment]
Note for incorporating with https://github.com/singularity-data/risingwave/pull/2387:

Semi and anti join will never produce out chunks with greater capacity than input chunks. So it could probably be optimized not to check the yield condition. Might be premature optimization.

File: src/stream/src/common/builder.rs [hunk 3]
```diff
@@ -122,34 +138,22 @@ impl StreamChunkBuilder {
...
-        let new_columns = new_arrays
-            .into_iter()
-            .map(|array_impl| Column::new(Arc::new(array_impl)))
-            .collect::<Vec<_>>();
 
-        Ok(StreamChunk::new(self.ops, new_columns, None))
+        self.inc_size()
     }
 
     pub fn take(&mut self) -> Result<Option<StreamChunk>> {
```
Comment (BugenZhao) [inline src/stream/src/common/builder.rs:159]: Should this method be `pub` now?

Comment (pleiadesian) [reply src/stream/src/common/builder.rs:159]: `pub take()` is called by HashJoinExecutor to generate a chunk from the remainder in the chunk builder.
https://github.com/singularity-data/risingwave/blob/35093ccddac4d688092a6162b780bdbac4a6adb7/src/stream/src/executor/hash_join.rs#L573

**Comment by yuhao-su** [comment]
The hash join is refactored with significant changes. Maybe I can help resolve the conflict?

**Comment by jon-chuang** [comment]
Hi @yuhao-su are you currently working on this? If not I can resolve the conflicts so as to get https://github.com/singularity-data/risingwave/issues/2428 unblocked?

**Comment by jon-chuang** [comment]
The conflicts are resolved! Could we get this reviewed again?

**Comment by yuhao-su** [comment]
LGTM!

**Review by yuhao-su — APPROVED** [review]

File: src/stream/src/executor/hash_join.rs [hunk 4]
```diff
@@ -226,55 +227,82 @@ struct HashJoinChunkBuilder<const T: JoinTypePrimitive, const SIDE: SideTypePrim
...
         // Outer sides
         } else if matched_row.is_zero_degree() && outer_side_null(T, SIDE) {
             // if the matched_row does not have any current
             // matches
+            if self
+                .stream_chunk_builder
+                .append_row_matched(Op::UpdateDelete, &matched_row.row)?
+                .is_some()
+            {
+                return Err(internal_error("`Op::UpdateDelete` should not yield chunk"));
```
Comment (BugenZhao) [inline src/stream/src/executor/hash_join.rs:302]: Can we directly `assert` here?

Comment (skyzh) [reply src/stream/src/executor/hash_join.rs:302]: I'd be okay with any solution :)

**Comment by skyzh** [comment]
So let's get this merged? I guess @pleiadesian has resolved conflicts of this PR for days 🤣

**Pull request merged**